### PR TITLE
Connecting trustvault to MetaMask

### PR DESCRIPTION
**Pull request by fodder1997**
This PR is to allow MetaMask to Connect to a TrustVault Wallet.

There are a number of design choices, e.g. https://github.com/MetaMask/metamask-extension/issues/7278 that we've raised separately and made some decisions on in this PR.

We'd love to hear the feedback on this and appreciate the time taken to review. 

Thanks

Trustology Team

### CODE REVIEW

**Comment by fodder1997** [comment]
I see we have some CI build failures - we'll take a look at that

**Comment by tmashuang** [comment]
The branch seems to be suffering to an using packages that are no longer currently using e.g. `react-hyperscript`. We've removed that package in #7680.

**Comment by cjeria** [comment]
Hi @fodder1997 can you please update to the latest build w/packages and resubmit this PR? Thanks!

**Comment by whymarrh** [comment]
> resubmit this PR

That is, ping us @fodder1997 when this is ready for review again

**Comment by whymarrh** [comment]
Though, that said, @cjeria this looks like it is good for a review [(chrome build](https://124508-42009758-gh.circle-artifacts.com/0/builds/metamask-chrome-7.8.0.zip), [firefox build](https://124508-42009758-gh.circle-artifacts.com/0/builds/metamask-firefox-7.8.0.zip)).

**Comment by fodder1997** [comment]
Thanks; we've updated to fix the issues you mentioned. Please shout if there's anything else.

File: app/manifest.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "__MSG_appName__",
   "short_name": "__MSG_appName__",
-  "version": "7.7.0",
+  "version": "7.8.0",
```
Comment (Gudahtt) [inline app/manifest.json]: Could you undo this change? We don't update the version in feature branches as a general rule. We prepare releases in a separate branch.

File: app/manifest.json [hunk 2]
```diff
@@ -55,8 +55,8 @@
   "content_scripts": [
     {
       "matches": [
-        "file://*/*",
-        "http://*/*",
+        "file://*/*",
```
Comment (Gudahtt) [inline app/manifest.json]: Nit: Theses extra spaces aren't necessary

File: package.json [hunk 3]
```diff
@@ -182,7 +182,8 @@
     "web3": "^0.20.7",
     "web3-stream-provider": "^4.0.0",
     "webrtc-adapter": "^6.3.0",
-    "xtend": "^4.0.1"
+    "xtend": "^4.0.1",
+    "graphql-request": "^1.8.2"
```
Comment (Gudahtt) [inline package.json]: Where is this dependency used?

**Review by Gudahtt — DISMISSED** [review]

File: package.json (hunk 3 above)
Comment (necati92) [reply package.json]: Thats an unused dependency will remove it.

File: app/manifest.json (hunk 1 above)
Comment (necati92) [reply app/manifest.json]: yeh sure

**Comment by necati92** [comment]
Hi MetaMask Team, would appreciate if we could get more feedback on this pull request so that we can move forward with it.  Many Thanks Nejati.

**Comment by whymarrh** [comment]
@fodder1997 @necati92 I'll merge in the latest `develop` to fix up some of the conflicts I've introduced and I'll give this a review

File: app/scripts/eth-trustvault-keyring.js [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+  _normalizeNum (buf) {
+    return new BigNumber(ethUtil.bufferToHex(buf)).toString(10)
+  }
+
+  _timeout (ms) {
+    return new Promise(resolve => setTimeout(resolve, ms))
+  }
+}
+
+TrustvaultKeyring.type = type
```
Comment (whymarrh) [inline app/scripts/eth-trustvault-keyring.js]: Can we move this onto the class as a `static` field? We can also avoid the constant if we reference it through `TrustvaultKeyring.type`.

File: app/scripts/eth-trustvault-keyring.js [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+module.exports = TrustvaultKeyring
```
Comment (whymarrh) [inline app/scripts/eth-trustvault-keyring.js]: Can we switch this to use ESM?

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -717,10 +726,11 @@ export default class MetamaskController extends EventEmitter {
     const hdKeyring = this.keyringController.getKeyringsByType('HD Key Tree')[0]
     const hdAccounts = await hdKeyring.getAccounts()
     const accounts = {
-      hd: hdAccounts.filter((item, pos) => (hdAccounts.indexOf(item) === pos)).map(address => ethUtil.toChecksumAddress(address)),
+      hd: hdAccounts.filter((item, pos) => hdAccounts.indexOf(item) === pos).map(address => ethUtil.toChecksumAddress(address)),
```
Comment (whymarrh) [inline app/scripts/metamask-controller.js]: Nit: can we drop these unrelated changes?

File: app/scripts/metamask-controller.js [hunk 7]
```diff
@@ -797,10 +807,10 @@ export default class MetamaskController extends EventEmitter {
   }
 
   //
-  // Hardware
+  // Hardware / Software
```
Comment (whymarrh) [inline app/scripts/metamask-controller.js]: We can just delete this comment

File: test/unit/app/controllers/metamask-controller-test.js [hunk 8]
```diff
@@ -898,7 +907,7 @@ describe('MetaMaskController', function () {
       })
 
       assert.deepEqual(addAddresses.args, [[['0x1', '0x2']]])
-      assert.deepEqual(syncWithAddresses.args, [[['0x1', '0x2']]])
+      assert.deepEqual(syncWithAddresses.args, [[['0x1', '0x2'] ]])
```
Comment (whymarrh) [inline test/unit/app/controllers/metamask-controller-test.js]: Nit: can we drop these unrelated changes?

File: test/unit/trustvaultKeyring/eth-trustvault-keyring-test.js [hunk 9]
```diff
@@ -0,0 +1,471 @@
+const TrustvaultKeyring = require('../../../app/scripts/eth-trustvault-keyring')
```
Comment (whymarrh) [inline test/unit/trustvaultKeyring/eth-trustvault-keyring-test.js]: Can we switch these to use ESM?

File: ui/app/components/app/account-menu/account-menu.component.js [hunk 10]
```diff
@@ -195,7 +195,7 @@ export default class AccountMenu extends Component {
     const { t } = this.context
     // Any account that's not from the HD wallet Keyring can be removed
     const { type } = keyring
-    const isRemovable = type !== 'HD Key Tree'
+    const isRemovable = !(type === 'HD Key Tree' || type === 'TrustVault')
```
Comment (whymarrh) [inline ui/app/components/app/account-menu/account-menu.component.js]: Nit: this suggestion here might be a cleaner diff

```suggestion
    const isRemovable = type !== 'HD Key Tree' && type !== 'TrustVault'
```

File: ui/app/ducks/app/app.js [hunk 11]
```diff
@@ -362,6 +362,11 @@ export default function reduceApp (state, action) {
         defaultHdPaths: newDefaults,
       })
 
+    case actions.TRUSTVAULT_PIN_CHALLENGE:
+      return extend(appState, {
```
Comment (whymarrh) [inline ui/app/ducks/app/app.js]: Nit: can we use the pattern introduced in #7795:

```js
case actions.TRUSTVAULT_PIN_CHALLENGE:
  return {
    ...appState,
    trustVault: { ...appState.trustVault, pinChallenge: action.value },
  }
```

File: ui/app/pages/create-account/connect-custodial/trustvault/components/circular-input-field/circular-input-field.component.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
+import React, { PureComponent } from 'react'
+import PropTypes from 'prop-types'
+import classnames from 'classnames'
+
+const CLASSNAME_CIRCULAR_INPUT = 'circular-input'
+const CLASSNAME_INLINE_BLOCK = 'inline-block-child'
+const CLASSNAME_BACKGROUND_FILLED = 'background-filled'
+const CLASSNAME_BACKGROUND_EMPTY = 'background-empty'
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/components/circular-input-field/circular-input-field.component.js]: All of these constants are used once? Can we inline them?

File: ui/app/pages/create-account/connect-custodial/trustvault/components/circular-input-field/circular-input-field.component.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+  render () {
+    return (
+      <div>
+        <input
+          className={classnames(
+            'circle-input',
+            CLASSNAME_CIRCULAR_INPUT,
+            CLASSNAME_INLINE_BLOCK,
+            this.props.filled()
+              ? CLASSNAME_BACKGROUND_FILLED
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/components/circular-input-field/circular-input-field.component.js]: `classnames` can handle this conditional case

File: ui/app/pages/create-account/connect-custodial/trustvault/components/circular-input-field/circular-input-field.component.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+              : CLASSNAME_BACKGROUND_EMPTY
+          )}
+          maxLength={this.props.maxLength.toString()}
+          type="password"
+          onChange={this.props.handleChange}
+          autoFocus={this.props.autoFocus}
+          ref={this.props.setRef}
+          onKeyPress={this.props.keyPress}
+        >
+        </input>
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/components/circular-input-field/circular-input-field.component.js]: Nit: can we collapse this empty tag?

File: ui/app/pages/create-account/connect-custodial/trustvault/components/circular-input-field/index.js [hunk 15]
```diff
@@ -0,0 +1,2 @@
+import CircularInputField from './circular-input-field.component'
+module.exports = CircularInputField
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/components/circular-input-field/index.js]: Can we keep file as just ESM?

File: ui/app/pages/create-account/connect-custodial/trustvault/components/star-placeholder/index.js [hunk 16]
```diff
@@ -0,0 +1,2 @@
+import StarPlaceholder from './star-placeholder.component'
+module.exports = StarPlaceholder
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/components/star-placeholder/index.js]: Can we keep file as just ESM?

File: ui/app/pages/create-account/connect-custodial/trustvault/components/star-placeholder/star-placeholder.component.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+import classnames from 'classnames'
+
+const CLASSNAME_STAR_PLACEHOLDER = 'star-placeholder'
+const CLASSNAME_INLINE_BLOCK = 'inline-block-child'
+
+export default class StarPlaceholder extends PureComponent {
+  render () {
+    return (
+      <span
+        className={classnames(CLASSNAME_STAR_PLACEHOLDER, CLASSNAME_INLINE_BLOCK)}
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/components/star-placeholder/star-placeholder.component.js]: These constants and this usage of classnames are all unnecessary, can we inline this `className` attr?

File: ui/app/pages/create-account/connect-custodial/trustvault/components/star-placeholder/star-placeholder.component.js [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+      >
+      ﹡
+      </span>
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/components/star-placeholder/star-placeholder.component.js]: All this JSX can fit on one line

File: ui/app/pages/create-account/connect-custodial/trustvault/components/star-placeholder/star-placeholder.component.js [hunk 19]
```diff
@@ -0,0 +1,17 @@
+import React, { PureComponent } from 'react'
+import classnames from 'classnames'
+
+const CLASSNAME_STAR_PLACEHOLDER = 'star-placeholder'
+const CLASSNAME_INLINE_BLOCK = 'inline-block-child'
+
+export default class StarPlaceholder extends PureComponent {
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/components/star-placeholder/star-placeholder.component.js]: Nit: can we make this a functional component

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+import TextField from '../../../../../components/ui/text-field'
+import { CONNECT_HARDWARE_ROUTE } from '../../../../../helpers/constants/routes'
+const ENTER_KEY = 13
+class EmailScreen extends PureComponent {
+  constructor (props) {
+    super(props)
+    this.state = {
+      email: null,
+    }
+    this.history = this.props.history
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: This class field seems unnecessary, esp. since it's used once? Can we inline this?

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 21]
```diff
@@ -0,0 +1,187 @@
+import React, { PureComponent } from 'react'
+const PropTypes = require('prop-types')
+import Button from '../../../../../components/ui/button'
+import TextField from '../../../../../components/ui/text-field'
+import { CONNECT_HARDWARE_ROUTE } from '../../../../../helpers/constants/routes'
+const ENTER_KEY = 13
+class EmailScreen extends PureComponent {
+  constructor (props) {
+    super(props)
+    this.state = {
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: Can we move this `state` field onto the class and remove the (resulting) empty constructor?

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+          largeLabel
+          fullWidth
+          onChange={event => this.setState({ email: event.target.value })}
+          onKeyDown={event => {
+            if (event.keyCode === ENTER_KEY) {
+              this.props.getTrustVaultPinChallenge(this.state.email)
+            }
+          }}
+        >
+        </TextField>
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: Nit: can we collapse this empty tag?

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+
+      </div>)
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: Nit: this closing paren should be on a newline

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+  }
+  renderTrustVaultInfoBox () {
+    return (
+      <div className="sw-connect__info-box">
+        <img className="sw-connect__info-box__info-icon" src="images/tvInfo.png" >
+
+        </img>
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: Nit: can we collapse this empty tag?

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+    )
+  }
+
+  renderTrustVaultLogo () {
+    return (
+      <div className="sw-connect__trustvault-logo">
+        <img className="sw-connect__trustvault-logo__img" src="images/trustvault-logo.png">
+
+
+        </img>
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: Nit: can we collapse this empty tag?

File: ui/app/pages/create-account/connect-custodial/trustvault/email/index.js [hunk 26]
```diff
@@ -0,0 +1,66 @@
+import React, { PureComponent } from 'react'
+const PropTypes = require('prop-types')
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/index.js]: Can we keep the imports here consistent, can we have this all be ESM?

File: ui/app/pages/create-account/connect-custodial/trustvault/email/index.js [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+const { connect } = require('react-redux')
+const actions = require('../../../../../store/actions')
+const EmailScreen = require('./email-screen')
+
+class ConnectTrustVaultEmailForm extends PureComponent {
+  constructor (props) {
+    super(props)
+    this.state = {
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/index.js]: Can we move this `state` field onto the class and remove the (resulting) empty constructor?

File: ui/app/pages/create-account/connect-custodial/trustvault/email/index.js [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+  }
+  renderContent () {
+    return (
+      <EmailScreen
+        history={this.props.history}
+        browserSupported={this.state.browserSupported}
+        getTrustVaultPinChallenge={this.props.getTrustVaultPinChallenge}
+      >
+
+      </EmailScreen>)
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/index.js]: Nit: collapse empty tag, newline for paren

File: ui/app/pages/create-account/connect-custodial/trustvault/email/index.js [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+const mapDispatchToProps = dispatch => ({
+  showAlert: msg => dispatch(actions.showAlert(msg)),
+  hideAlert: () => dispatch(actions.hideAlert()),
+})
+
+ConnectTrustVaultEmailForm.contextTypes = {
+  metricsEvent: PropTypes.func,
+}
+
+module.exports = connect(
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/index.js]: Can we use ESM here?

File: ui/app/pages/create-account/create-account.component.js [hunk 30]
```diff
@@ -76,3 +87,5 @@ CreateAccountPage.propTypes = {
 CreateAccountPage.contextTypes = {
   t: PropTypes.func,
 }
+
+
```
Comment (whymarrh) [inline ui/app/pages/create-account/create-account.component.js]: Nit: these newlines are unnecessary

File: ui/app/pages/create-account/connect-custodial/trustvault/pin/pin-screen.js [hunk 31]
```diff
@@ -0,0 +1,221 @@
...
+  submitTrustVaultPinChallenge: PropTypes.func.isRequired,
+  onCancelLogin: PropTypes.func.isRequired,
+  email: PropTypes.string,
+}
+
+PinScreen.contextTypes = {
+  t: PropTypes.func,
+}
+
+module.exports = PinScreen
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/pin/pin-screen.js]: Can we use ESM here?

File: ui/app/pages/create-account/connect-custodial/trustvault/index.js [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+  render = () => {
+    return (
+      <div className="new-account__header">
+        <div className="trustvault-connect">
+          {this.renderError()}
+          {!this.state.pinChallenge ? this.renderEmailForm() : this.renderPinForm()}
+          <Switch>
+            <Route exact path={TRUSTVAULT_EMAIL_ROUTE} component={ConnectTrustVaultEmailForm} >
+
+            </Route>
```
Comment (whymarrh) [inline ui/app/pages/create-account/connect-custodial/trustvault/index.js]: Nit: can we collapse this/these empty tags?

File: test/unit/trustvaultKeyring/eth-trustvault-keyring-test.js [hunk 33]
```diff
@@ -0,0 +1,471 @@
...
+        .then(result => assert.fail(`Expected to fail returned with result ${result}`))
+        .catch(e => {
+          assert.equal(e, 'Transaction cancelled by user')
+          done()
+        })
+    })
+    it('should return Error transaction if transaction status is ERROR', (done) => {
+      sandbox.stub(TrustvaultKeyring.prototype, '_request').resolves({ userWallet: { getTransactionInfo: { status: 'ERROR', signedTransaction: 'signedTransaction' } } })
+      const trustVaultKeyring = new TrustvaultKeyring({ auth: 'test' })
+      trustVaultKeyring._pollTransaction('txID')
```
Comment (whymarrh) [inline test/unit/trustvaultKeyring/eth-trustvault-keyring-test.js]: Can we use `async`/`await` for these new tests?

**Review by whymarrh — CHANGES_REQUESTED** [review]
@fodder1997 @necati92 I've completed a partial review of this but I'll stop here to avoid making too many redundant comments—can we clean up this implementation a bit before continuing with review? Admittedly this PR points out a good few places where our ESLint rules should be more strict, but in lieu of that, can you tidy up the code a bit.

A few themes:

1. Inconsistent usages of CommonJS and ESM in the fresh files—can we keep everything to ESM?
2. A good few unrelated changes here—I dropped a few but there's still a few instances of diff noise that's a bit distracting (e.g. empty lines)
3. There are a few (at least one?) instances of new components having two distinct sets of `propTypes` added to them, one in the class and one at the bottom of the file
4. Can we lean on [class fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_fields) where possible?

I can take another look/review this again after these low-hanging fruit have been addressed.

**Comment by necati92** [comment]
@whymarrh thanks for reviewing, we will address your  review comments and update the pr

**Comment by whymarrh** [comment]
Also, apologies for the lint failure in the merge commit, that'll need to be addressed as well

File: test/e2e/from-import-ui.spec.js [hunk 34]
```diff
@@ -335,9 +335,9 @@ describe('Using MetaMask with an existing account', function () {
     })
   })
 
-  describe('Connects to a Hardware wallet', () => {
-    it('choose Connect Hardware Wallet from the account menu', async () => {
-      const [connectAccount] = await driver.findElements(By.xpath(`//div[contains(text(), 'Connect Hardware Wallet')]`))
+  describe('Connects to an External  wallet', () => {
+    it('choose Connect External Wallet from the account menu', async () => {
+      const [connectAccount] = await driver.findElements(driver, By.xpath(`//div[contains(text(), 'Connect External Wallet')]`))
```
Comment (Gudahtt) [inline test/e2e/from-import-ui.spec.js]:
```suggestion
      const [connectAccount] = await driver.findElements(By.xpath(`//div[contains(text(), 'Connect External Wallet')]`))
```
The `driver` param isn't needed anymore, now that this is being called from the `driver`.

Comment (necati92) [reply test/e2e/from-import-ui.spec.js]: Thanks :), merged that by mistake, should be fixed now

Comment (Gudahtt) [reply test/e2e/from-import-ui.spec.js]: Great thanks!

**Comment by necati92** [comment]
Hey @whymarrh , we have made the suggested changes.

**Comment by necati92** [comment]
Thank you very much @whymarrh

File: app/scripts/eth-trustvault-keyring.js [hunk 35]
```diff
@@ -0,0 +1,445 @@
+const { EventEmitter } = require('events')
+const request = require('request-promise')
+const hdPathString = `m/44'/60'/0'/0`
+const ethUtil = require('ethereumjs-util')
+const type = 'TrustVault'
+const log = require('loglevel')
+const BigNumber = require('bignumber.js')
+const trustVaultBridgeUrl = 'https://t1tl7x0ii7.execute-api.eu-west-1.amazonaws.com/production/graphql'
+const apiKey = 'da2-ekziyghfhzda7hsapyziwdurwm'
```
[the remark this answers is not in the document]
Comment (whymarrh) [reply app/scripts/eth-trustvault-keyring.js]: @fodder1997? @necati92?

**Review by whymarrh — CHANGES_REQUESTED** [review]

**Comment by necati92** [comment]
Hi @whymarrh  that's our API Key and api url that we use to talk to our backend service in order to connect the wallet.They are not supposed to be secret.I hope that clarifies your question.

**Comment by necati92** [comment]
Hi MetaMask team, would really appreciate if we could get more feedback on the pull request please.

**Comment by whymarrh** [comment]
@necati92 apologies for the slow review, we are/I am juggling a lot of different work. Would you mind rebasing this on the latest `develop` an addressing the conflicts here. I've reviewed most of the implementation and it looks alright—@tmashuang would you mind taking this for a spin and seeing how this works?

**Comment by whymarrh** [comment]
We don't need to add `request-promise` here, we already have `node-fetch`/[`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)

**Comment by necati92** [comment]
Hey @whymarrh  really appreciate for you having a look. I have resolved the merge conflicts and  removed request-promise.  Please let us know if you have further questions or want us to walk trough how it works.

**Review by tmashuang — CHANGES_REQUESTED** [review]

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 36]
```diff
@@ -0,0 +1,160 @@
...
+  state = {
+    email: null,
+  }
+
+  renderBackButton () {
+    return (
+      <div className="sw-connect__back" onClick={(_) => this.props.history.push(CONNECT_HARDWARE_ROUTE)}>
+        <div className="sw-connect__list__back-caret" />
+        <div className="sw-connect__list__back-caret__back">{this.context.t('back')}</div>
+      </div>
```
Comment (tmashuang) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: Same comment, can we get a cursor pointer hover here as well?

File: ui/app/pages/create-account/connect-hardware/connect-screen.js [hunk 37]
```diff
@@ -111,16 +138,25 @@ class ConnectScreen extends Component {
     const links = {
       trezor: `<a class='hw-connect__get-hw__link' href='https://shop.trezor.io/?a=metamask' target='_blank'>Trezor</a>`,
       ledger: `<a class='hw-connect__get-hw__link' href='https://www.ledger.com/products/ledger-nano-s?r=17c4991a03fa&tracker=MY_TRACKER' target='_blank'>Ledger</a>`,
+      trustvault: `<a class='sw-connect__get-sw__link' href='https://trustology.io/get-started/'>here</a>`,
```
Comment (tmashuang) [inline ui/app/pages/create-account/connect-hardware/connect-screen.js]: Would like to have this link open in a new tab.

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 38]
```diff
@@ -0,0 +1,160 @@
...
+        />
+      </div>
+    )
+  }
+  renderTrustVaultInfoBox () {
+    return (
+      <div className="sw-connect__info-box">
+        <img className="sw-connect__info-box__info-icon" src="images/tvInfo.png" />
+        <div className="sw-connect__info-box__not-user">{this.context.t('trustVaultNotUser')}</div>
+        <div className="sw-connect__info-box__ios">{this.context.t('trustVaultIos')}</div>
```
Comment (tmashuang) [inline ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: This seems like it will not last very long, since the android app will come out and you can sign up via email. This can probably be omitted along with the associated locale message.

File: app/scripts/metamask-controller.js [hunk 39]
```diff
@@ -860,6 +877,23 @@ export default class MetamaskController extends EventEmitter {
     return accounts
   }
 
+  /**
+   * Fetch account list from a TrustVault wallet
+   *
+   * @returns [] accounts
+   */
+  async connectCustodialWallet (deviceName, auth, hdPath = null) {
+    await this.keyringController.removeEmptyKeyrings()
```
Comment (tmashuang) [inline app/scripts/metamask-controller.js]: Is this necessary?

File: ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js [hunk 40]
```diff
@@ -0,0 +1,160 @@
...
+        <div className="sw-connect__info-box__get-started">{this.context.t('trustVaultGetStarted')}</div>
+        <div
```
[the remark this answers is not in the document]
Comment (whymarrh) [reply ui/app/pages/create-account/connect-custodial/trustvault/email/email-screen.js]: Let's lean on an `a` element if possible

**Comment by necati92** [comment]
Hey MetaMask team, we pushed the changes last week.I would appreciate if we could get a review on them. Thanks Necati

File: app/scripts/metamask-controller.js (hunk 39 above)
Comment (tmashuang) [reply app/scripts/metamask-controller.js]: Omit this line.

**Comment by tmashuang** [comment]
There is a conflict in `app/scripts/metamask-controller.js` where we had to update and change the ledger keyring. Please update to resolve merge conflict.

**Comment by whymarrh** [comment]
> Hey MetaMask team, we pushed the changes last week.I would appreciate if we could get a review on them. Thanks Necati

@necati92 @fodder1997 just to follow-up, and apologies for the slow responses. There is a good bit of code here that we'll have to maintain going forward and we're already working on getting a large major version out the door ([`v7.7.8...develop`](https://github.com/MetaMask/metamask-extension/compare/v7.7.8...develop)). That is to say that this is, unfortunately, a lower priority for us right now—we'll be sure to update you here when this is back on our radar.

**Review by miranda76 — DISMISSED** [review]

**Comment by necati92** [comment]
Hey thank you very much for approving the pull request, 
There was merge conflicts on the branch, so I have fixed them.
I would appreciate If you could relook on the changes.
Many Thanks
Necati

**Comment by danjm** [comment]
@necati92 Sorry for the long delay in communication. We are not going to accept this PR at this time. However, in the future we will possibly be supporting APIs that allow such functionality to be added to metamask via our [snaps system](https://medium.com/metamask/introducing-the-next-evolution-of-the-web3-wallet-4abdf801a4ee).
